### PR TITLE
Convert Guest class to JS

### PR DESCRIPTION
**Pull request by robertknight**
This PR is a mostly straightforward conversion of `src/annotator/guest.coffee` to JS. I avoided converting the tests in this PR since the diff is quite large.

Testing of this PR mostly relies on the existing tests. Converting the code to JS surfaced the fact that there are ~40 lines in the `Guest` class which are missing tests. I may look at adding those as part of this PR.

**TODO**
- [x] Add additional tests for code that is not currently covered (Edit: I've decided to defer this until after the existing tests are converted to JS)
- [x] Do another review of the converted code and look for anything in the CoffeeScript that was not idiomatically converted to JS

### CODE REVIEW

File: src/annotator/guest.js [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+    }
+
+    const locate = target => {
+      // Check that the anchor has a TextQuoteSelector -- without a
+      // TextQuoteSelector we have no basis on which to verify that we have
+      // reanchored correctly and so we shouldn't even try.
+      //
+      // Returning an anchor without a range will result in this annotation being
+      // treated as an orphan (assuming no other targets anchor).
+      if (!(target.selector ?? []).some(s => s.type === 'TextQuoteSelector')) {
```
Comment (LMS007) [inline src/annotator/guest.js:374]: nullish op, thats new!

**Review by LMS007 — APPROVED** [review]
👍 
Very straightforward, just a few comments below, and please proceed with 2/2

File: src/annotator/guest.js [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+
+    this.crossframe.onConnect(() => this._setupInitialState(config));
+
+    // Whether clicks on non-highlighted text should close the sidebar
+    this.closeSidebarOnDocumentClick = true;
+    this._connectAnnotationSync();
+    this._connectAnnotationUISync(this.crossframe);
+
+    // Load plugins
+    for (let name of Object.keys(this.options || {})) {
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/guest.js:142]: Indeed, you're right. The `|| {}` shouldn't be needed.

File: src/annotator/guest.js [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+      this.setVisibleHighlights(state);
+    });
+  }
+
+  destroy() {
+    this._removeElementEvents();
+    this.selections.unsubscribe();
+    this.adder.remove();
+
+    this.element.find('.hypothesis-highlight').each(function () {
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/guest.js:328]: The callback references `this` to refer to the current `.hypothesis-highlight` element in the collection. The callback is passed the element as an argument as well, so I'll change it to use that instead.

**Comment by robertknight** [comment]
The `Guest` class does have several gaps in its test coverage which are now visible as a result of this PR that need to be fixed. However I think to avoid conflicts with other developers and to avoid scope creep in this PR I think it makes sense to land those as is and then plug the gaps afterwards. For now I have gone through the functionality that codecov has highlighted as not been covered by tests and checked it manually.

File: src/annotator/guest.js [hunk 4]
```diff
@@ -0,0 +1,660 @@
+import CustomEvent from 'custom-event';
+import $ from 'jquery';
+import scrollIntoView from 'scroll-into-view';
+
+import Delegator from './delegator';
+import * as adder from './adder';
```
Comment (esanzgar) [inline src/annotator/guest.js:6]: Is it better to be more explicit? 

`import {Adder} from './adder';

File: src/annotator/guest.js [hunk 5]
```diff
@@ -0,0 +1,660 @@
...
+    }
+
+    // Setup event handlers on the root element
+    this._elementEventListeners = [];
+    this._setupElementEvents();
+  }
+
+  // Add DOM event listeners for clicks, taps etc. on the document and
+  // highlights.
+  _setupElementEvents() {
```
[the remark this answers is not in the document]
Comment (esanzgar) [reply src/annotator/guest.js:156]: In ES2019 added private class fields:
`#elementEventListeners = [];  // this is private`
`this.#elementEventListener = []`

There is a proposal to extend it for class methods.

**Pull request merged**